### PR TITLE
WAITP-1504: Datatrak reports

### PR DESCRIPTION
**Pull request by acdunham**
### Changes:
- Added reports view to Datatrak Web

---

### Screenshots:
Screenshot 2023-12-15 at 11 02 07 am

Screenshot 2023-12-15 at 11 02 37 am
Screenshot 2023-12-15 at 11 03 33 am

Screenshot 2023-12-15 at 11 04 12 am

### CODE REVIEW

File: packages/datatrak-web/src/__tests__/features/Reports/Reports.test.tsx [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+  rest.get('*/v1/surveys', (_, res, ctx) => {
+    return res(ctx.status(200), ctx.json(surveysData));
+  }),
+);
+
+beforeAll(() => server.listen());
+afterEach(() => server.resetHandlers());
+afterAll(() => server.close());
+
+describe('Reports', () => {
```
Comment (tcaiger) [inline packages/datatrak-web/src/__tests__/features/Reports/Reports.test.tsx:66]: It's great to see these front end tests coming through

File: packages/datatrak-web/src/components/PageTitleBar.tsx [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+`;
+
+interface PageTitleBarProps {
+  title?: ReactNode;
+  children?: ReactNode;
+  isTransparent?: boolean;
+  Icon: typeof SvgIcon;
+}
+
+export const PageTitleBar = ({ isTransparent, title, children, Icon }: PageTitleBarProps) => {
```
Comment (tcaiger) [inline packages/datatrak-web/src/components/PageTitleBar.tsx:47]: Nice refactor to extract this out

File: packages/datatrak-web/src/views/ReportsPage.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  text-decoration: underline;
+`;
+
+export const ReportsPage = () => {
+  return (
+    <PageContainer>
+      <PageTitleBar title="Reports" Icon={ReportsIcon} isTransparent />
+      <Wrapper>
+        <Container>
+          <Typography>
```
Comment (tcaiger) [inline packages/datatrak-web/src/views/ReportsPage.tsx:41]: Looks like the line height needs to be upped

File: packages/datatrak-web/src/views/ReportsPage.tsx [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  }
+`;
+
+const InlineLink = styled(Link)`
+  text-decoration: underline;
+`;
+
+export const ReportsPage = () => {
+  return (
+    <PageContainer>
```
Comment (tcaiger) [inline packages/datatrak-web/src/views/ReportsPage.tsx:37]: Looks like the border radius needs to be upped

File: packages/datatrak-web/src/views/AccountSettingsPage/ChangePasswordSection/ChangePasswordForm.tsx [hunk 5]
```diff
@@ -95,7 +94,7 @@ export const ChangePasswordForm = () => {
   const submissionShouldBeDisabled = isValidating || !isValid || isSubmitting;
 
   return (
-    <StyledForm onSubmit={handleSubmit(attemptPasswordChange)} formContext={formContext}>
```
Comment (tcaiger) [inline packages/datatrak-web/src/views/AccountSettingsPage/ChangePasswordSection/ChangePasswordForm.tsx:98]: Is this change intentional? Seems like it will break the change password form?

Comment (acdunham) [reply packages/datatrak-web/src/views/AccountSettingsPage/ChangePasswordSection/ChangePasswordForm.tsx:98]: Yep - that reusable component in `ui-components` does the wrapping of the `onSubmit` handler for us

File: packages/datatrak-web/src/routes/Routes.tsx [hunk 6]
```diff
@@ -97,6 +109,14 @@ export const Routes = () => {
             <Route path={ROUTES.VERIFY_EMAIL_RESEND} element={<VerifyEmailResendPage />} />
           </Route>
         </Route>
+        <Route
```
Comment (tcaiger) [inline packages/datatrak-web/src/routes/Routes.tsx:112]: I think it would be nice to re-organise this a bit. Could we move it into the private routes section above so that we don't need to wrap it in the private route.

File: packages/datatrak-web/src/routes/Routes.tsx [hunk 7]
```diff
@@ -37,6 +39,16 @@ const LoggedInRedirect = ({ children }) => {
   return <Navigate to={user.projectId ? ROUTES.HOME : ROUTES.PROJECT_SELECT} replace={true} />;
 };
 
+const ReportsRedirect = () => {
+  const { isLoading, data: user } = useUser();
```
Comment (tcaiger) [inline packages/datatrak-web/src/routes/Routes.tsx:43]:
```suggestion
  const { isLoading, data: user } = useCurrentUser();
```
If we used useCurrentUser then we wouldn't have to worry about loading state but even better would be to either combine this check into the Private Route as some kind of Admin check or extend the private route for Admin checking and then remove the ReportsRedirect component entirely.

File: packages/datatrak-web/src/features/Reports/Reports.tsx [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+          <EmailDownloadAlert>
+            This export is taking a while, and will continue in the background. You will be emailed
+            when the export process completes.
+          </EmailDownloadAlert>
+        )}
+        <SurveysInput />
+        <EntityLevelInput />
+        {/** Render both of these, and then handle hiding with isActive prop. This means that when we change entityLevel we don't get weird rendering of selected values from the wrong entity type */}
+        <EntitySelectorInput entityLevel="country" />
+        <EntitySelectorInput entityLevel="entity" />
```
Comment (tcaiger) [inline packages/datatrak-web/src/features/Reports/Reports.tsx:135]:
```suggestion
                <EntitySelectorInput key={selectedEntityLevel} selectedEntityLevel={selectedEntityLevel} />
```

File: packages/datatrak-web/src/features/Reports/Reports.tsx [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+      <Form onSubmit={handleSubmit(onSubmit)} formContext={formContext} $isLoading={isLoading}>
+        {responseData?.emailTimeoutHit && (
+          <EmailDownloadAlert>
+            This export is taking a while, and will continue in the background. You will be emailed
+            when the export process completes.
+          </EmailDownloadAlert>
+        )}
+        <SurveysInput />
+        <EntityLevelInput />
+        {/** Render both of these, and then handle hiding with isActive prop. This means that when we change entityLevel we don't get weird rendering of selected values from the wrong entity type */}
```
Comment (tcaiger) [inline packages/datatrak-web/src/features/Reports/Reports.tsx:133]: hmmm this work around is ok but I think the problem could more elegantly be solved by using a render key which is what the react.dev docs recommend [here](https://react.dev/learn/you-might-not-need-an-effect#resetting-all-state-when-a-prop-changes). Then we wouldn't need the isActive code which would be nice. It would mean re-jigging the defaultValues and onSubmit too so if that leads to nasty problems then I'm happy to leave this it as is.

**Review by tcaiger — CHANGES_REQUESTED** [review]
Looks really good overall. I just left a couple of suggested refactors to look.

File: packages/datatrak-web/src/routes/PrivateRoute.tsx [hunk 10]
```diff
@@ -39,5 +39,12 @@ export const PrivateRoute = ({ children }: { children?: ReactNode }): any => {
         }}
       />
     );
+
+  const isAdminOnlyRoute = ADMIN_ONLY_ROUTES.includes(location.pathname);
+
+  // If the user is logged in but is not an admin and is trying to access an admin only page, redirect to the home page
+  if (isAdminOnlyRoute && !hasAdminPanelAccess) {
```
Comment (tcaiger) [inline packages/datatrak-web/src/routes/PrivateRoute.tsx:46]: This is just lovely if I don't way so myself

**Review by tcaiger — APPROVED** [review]
